src/core/service.c

```diff
@@ -1561,21 +1587,49 @@ static int service_spawn(
                 }
         }
 
-        if (flags & EXEC_SETENV_RESULT) {
-                if (asprintf(our_env + n_env++, "SERVICE_RESULT=%s", service_result_to_string(s->result)) < 0)
-                        return -ENOMEM;
+        if ((flags & EXEC_SETENV_RESULT) ^ (flags & EXEC_SETENV_MONITOR_RESULT)) {
+                // We don't expect both EXEC_SETENV_RESULT and EXEC_SETENV_MONITOR_RESULT to be set.
+                Service *env_source;
 
-                if (s->main_exec_status.pid > 0 &&
-                    dual_timestamp_is_set(&s->main_exec_status.exit_timestamp)) {
-                        if (asprintf(our_env + n_env++, "EXIT_CODE=%s", sigchld_code_to_string(s->main_exec_status.code)) < 0)
-                                return -ENOMEM;
+                if (flags & EXEC_SETENV_MONITOR_RESULT)
+                        env_source = service_get_triggering_service(s);
+                else
+                        env_source = s;
 
-                        if (s->main_exec_status.code == CLD_EXITED)
-                                r = asprintf(our_env + n_env++, "EXIT_STATUS=%i", s->main_exec_status.status);
-                        else
-                                r = asprintf(our_env + n_env++, "EXIT_STATUS=%s", signal_to_string(s->main_exec_status.status));
-                        if (r < 0)
+                if (env_source) {
+                        if (asprintf(our_env + n_env++, "%sSERVICE_RESULT=%s",
+                                     flags & EXEC_SETENV_MONITOR_RESULT ? "MONITOR_" : "",
```

Comment:
define a const variable `monitor_prefix` or so and reuse it to make it shorter and more readable?

Reply:
done.